internal/pkg/agent/cmd/watch.go

```diff
@@ -107,10 +108,35 @@ func watchCmd(log *logp.Logger, cfg *configuration.Configuration) error {
 	errorCheckInterval := cfg.Settings.Upgrade.Watcher.ErrorCheck.Interval
 	ctx := context.Background()
 	if err := watch(ctx, tilGrace, errorCheckInterval, log); err != nil {
-		log.Error("Error detected proceeding to rollback: %v", err)
+		log.Error("Error detected, proceeding to rollback: %v", err)
+
+		// If we are upgrading from version >= 8.12.0, marker.Details should be non-nil
+		// because the Agent we upgraded FROM would've written upgrade details in the upgrade
+		// marker. However, if we're upgrading from version < 8.12.0, the marker won't
+		// contain upgrade details, so we populate them now.
+		if marker.Details == nil {
+			actionID := ""
+			if marker.Action != nil {
+				actionID = marker.Action.ActionID
+			}
+			marker.Details = details.NewDetails(version.GetAgentPackageVersion(), details.StateRollback, actionID)
+		}
+
+		marker.Details.SetState(details.StateRollback)
+		err = upgrade.SaveMarker(marker)
```

Comment:
i wonder if we want to guard this with some retry. we're accessing marker file from 2 processes now and when one process opens marker even for reading the other may fail on windows

Reply:
Thanks, I was not aware this was an issue on Windows.  So it seems it would make sense to add retry in `upgrade.SaveMarker` and `upgrade.LoadMarker` both.  I'll make the change.

Reply:
Implemented in c174487f9a.  Let me know what you think.

Reply:
That this can fail makes detecting the upgrade rollback state best effort. We might be able to improve this, and make it so that only getting the reason for a rollback is best effort.

When the agent first starts up it reads the upgrade marker to determine if it needs to acknowledge a pending upgrade action. This eventually ends up here:

https://github.com/elastic/elastic-agent/blob/f7dcbd74ce0e27d3b46b478755168c66d78dc111/internal/pkg/agent/application/upgrade/upgrade.go#L210-L236

If the agent at this point sees that there is an upgrade marker file, and the target version does not match the agent's current version, that should be enough to conclude the upgrade was rolled back. I don't think there's another situation that leads to this.

This should guarantee we can report rollbacks if this write fails for whatever reason, but we just won't know the reason why.

This may be complicated by the fact that it looks like the upgrade watcher sometimes removes the marker file itself? I'm not sure why or if this is correct because per the code above the presence of the marker is how the agent knows to ack the upgrade. This could be causing some of our "stuck in updating" problems if I am following this correctly.

https://github.com/elastic/elastic-agent/blob/f7dcbd74ce0e27d3b46b478755168c66d78dc111/internal/pkg/agent/cmd/watch.go#L118-L128

Reply:
> If the agent at this point sees that there is an upgrade marker file, and the target version does not match the agent's current version, that should be enough to conclude the upgrade was rolled back.

I think the logic here needs to be revised slightly. 

Older versions of Agent that don't contain upgrade details in the Upgrade Marker will only contain the version that Agent was being upgraded _from_ in the Upgrade Marker (the `prevVersion` field), not the target version of the upgrade.

As such, I think the condition indicating a rollback is if the `prevVersion` value from the Upgrade Marker matches the agent's current version.

---

Also, while I totally agree with this idea of doing a best effort under the said circumstances of recording in upgrade details that we've rolled back, I don't think this `Ack()` method is the right place to do it as it's specific to Fleet and also it's concern is acknowledging the upgrade.  I think we can make this logic the responsibility of the Upgrade Marker Watcher introduced in this PR as that's not a Fleet-specific component.

Reply:
Implemented in 210af5c9a5.  Please let me know what you think.